src/set_ref.rs

```diff
@@ -0,0 +1,229 @@
+use crate::iter::*;
+use crate::{HashSet, GuardRef, TryInsertError};
+use crossbeam_epoch::Guard;
+use std::borrow::Borrow;
+use std::fmt::{self, Debug, Formatter};
+use std::hash::{BuildHasher, Hash};
+use std::ops::Index;
+
+/// A reference to a [`HashSet`], constructed with [`HashSet::pin`] or [`HashSet::with_guard`].
+///
+/// The current thread will be pinned for the duration of this reference.
+/// Keep in mind that this prevents the collection of garbage generated by the set.
+pub struct HashSetRef<'set, T, S = crate::DefaultHashBuilder> {
+    set: &'set HashSet<T, S>,
+    guard: GuardRef<'set>,
+}
+
+impl<K, V, S> HashSet<T, S> {
+    /// Get a reference to this set with the current thread pinned.
+    ///
+    /// Keep in mind that for as long as you hold onto this, you are preventing the collection of
+    /// garbage generated by the set.
+    pub fn pin(&self) -> HashSetRef<'_, T, S> {
+        HashSetRef {
+            guard: GuardRef::Owned(self.guard()),
+            set: &self,
+        }
+    }
+
+    /// Get a reference to this set with the given guard.
+    pub fn with_guard<'g>(&'g self, guard: &'g Guard) -> HashSetRef<'g, T, S> {
+        HashSetRef {
+            set: &self,
+            guard: GuardRef::Ref(guard),
+        }
+    }
+}
+
+impl<T, S> HashSetRef<'_, T, S> {
+    /// An iterator visiting all key-value pairs in arbitrary order.
+    /// The iterator element type is `(&'g K, &'g V)`.
+    /// See also [`HashSet::iter`].
+    pub fn iter(&self) -> Keys<'_, T, ()> {
+        self.set.iter(&self.guard)
+    }
+
+    /// Returns the number of entries in the set.
+    /// See also [`HashSet::len`].
+    pub fn len(&self) -> usize {
+        self.set.len()
+    }
+
+    /// Returns `true` if the set is empty. Otherwise returns `false`.
+    /// See also [`HashSet::is_empty`].
+    pub fn is_empty(&self) -> bool {
+        self.set.is_empty()
+    }
+}
+
+impl<K, V, S> HashSetRef<'_, K, V, S>
+where
+    K: Clone,
+{
+    /// Tries to reserve capacity for at least additional more elements.
+    /// See also [`HashSet::reserve`].
+    pub fn reserve(&self, additional: usize) {
+        self.set.reserve(additional, &self.guard)
+    }
+
+    /// Removes all entries from this set.
+    /// See also [`HashSet::clear`].
+    pub fn clear(&self) {
+        self.set.clear(&self.guard);
+    }
+}
+
+impl<K, V, S> HashSetRef<'_, K, V, S>
+where
+    K: Hash + Eq,
+    S: BuildHasher,
+{
+    /// Tests if `key` is a key in this table.
+    /// See also [`HashSet::contains_key`].
+    pub fn contains<Q>(&self, value: &Q) -> bool
```

Comment:
I think most (all?) of the `HashSet` docs refer to the items in the set as `values` rather than `keys`. Maybe we should update these to be consistent with that?

Reply:
I haven't paid much attention to the docs yet. I'm trying to make everything compile before that. Still, I'll work on this.

Reply:
Just fixed it in 7bc63b0.